diff -Naur mysql-5.5.7-rc/include/mysql/client_plugin.h mysql-5.5.7-rc-patch//include/mysql/client_plugin.h
--- mysql-5.5.7-rc/include/mysql/client_plugin.h	2010-11-02 22:01:11.000000000 +0000
+++ mysql-5.5.7-rc-patch//include/mysql/client_plugin.h	2010-12-03 03:46:30.323659543 +0000
@@ -23,8 +23,10 @@
 */
 #define MYSQL_CLIENT_PLUGIN_INCLUDED
 
+#if !defined(MYSQL_ABI_CHECK)
 #include <stdarg.h>
 #include <stdlib.h>
+#endif
 
 /* known plugin types */
 #define MYSQL_CLIENT_reserved1               0
diff -Naur mysql-5.5.7-rc/include/mysql/client_plugin.h.pp mysql-5.5.7-rc-patch//include/mysql/client_plugin.h.pp
--- mysql-5.5.7-rc/include/mysql/client_plugin.h.pp	2010-11-02 22:01:11.000000000 +0000
+++ mysql-5.5.7-rc-patch//include/mysql/client_plugin.h.pp	2010-12-03 03:46:00.313673730 +0000
@@ -1,5 +1,3 @@
-#include <stdarg.h>
-#include <stdlib.h>
 struct st_mysql_client_plugin
 {
   int type; unsigned int interface_version; const char *name; const char *author; const char *desc; unsigned int version[3]; const char *license; void *mysql_api; int (*init)(char *, size_t, int, va_list); int (*deinit)(); int (*options)(const char *option, const void *);
