=== modified file 'kewpie.py'
--- old/kewpie.py	2012-01-09 21:17:09 +0000
+++ new/kewpie.py	2012-01-20 17:37:28 +0000
@@ -49,9 +49,9 @@ from lib.test_mgmt.execution_management 
 # We base / look for a lot of things based on the location of
 # the kewpie.py file
 qp_rootdir = os.path.dirname(os.path.abspath(sys.argv[0]))
-#project_name = 'percona-xtradb-cluster'
+project_name = 'percona-xtradb-cluster'
 #project_name = 'xtrabackup'
-project_name = None
+#project_name = None
 defaults = get_defaults(qp_rootdir,project_name)
 variables = test_run_options.parse_qp_options(defaults)
 variables['qp_root'] = qp_rootdir

